core/src/main/java/com/linecorp/armeria/common/HttpHeadersBase.java

```diff
@@ -0,0 +1,1214 @@
+        }
+    }
+
+    private int index(int hash) {
+        return hash & hashMask;
+    }
+
+    private void add0(int h, int i, AsciiString name, CharSequence value) {
+        ensureMutability();
+        // Update the hash table.
+        entries[i] = new HeaderEntry(h, name, value, entries[i]);
+        ++size;
+    }
+
+    private boolean addFast(
+            Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
+        if (!(headers instanceof HttpHeadersBase)) {
+            return false;
+        }
+
+        final HttpHeadersBase headersBase = (HttpHeadersBase) headers;
+        HeaderEntry e = headersBase.head.after;
+        while (e != headersBase.head) {
+            final AsciiString key = e.key;
+            final CharSequence value = e.value;
+            assert key != null;
+            assert value != null;
+            add0(e.hash, index(e.hash), key, value);
+            e = e.after;
+        }
+
+        return true;
+    }
+
+    private void addSlow(
+            Iterable<? extends Entry<? extends CharSequence, ? extends CharSequence>> headers) {
+        // Slow copy
+        for (Entry<? extends CharSequence, ? extends CharSequence> header : headers) {
+            add(header.getKey(), header.getValue());
+        }
+    }
+
+    private void addObjectSlow(
+            Iterable<? extends Entry<? extends CharSequence, ?>> headers) {
+        // Slow copy
+        for (Entry<? extends CharSequence, ?> header : headers) {
+            addObject(header.getKey(), header.getValue());
+        }
+    }
+
+    /**
+     * Removes all the entries whose hash code equals {@code h} and whose name is equal to {@code name}.
+     *
+     * @return the first value inserted, or {@code null} if there is no such header.
+     */
+    @Nullable
+    private CharSequence remove0(int h, int i, CharSequence name) {
+        ensureMutability();
+        HeaderEntry e = entries[i];
+        if (e == null) {
+            return null;
+        }
+
+        CharSequence value = null;
+        HeaderEntry next = e.next;
+        while (next != null) {
+            if (next.hash == h && keyEquals(next.key, name)) {
+                value = next.value;
+                e.next = next.next;
+                next.remove();
+                --size;
+            } else {
+                e = next;
+            }
+
+            next = e.next;
+        }
+
+        e = entries[i];
+        if (e.hash == h && keyEquals(e.key, name)) {
+            if (value == null) {
+                value = e.value;
+            }
+            entries[i] = e.next;
+            e.remove();
+            --size;
+        }
+
+        return value;
+    }
+
+    private void ensureMutability() {
+        checkState(!immutable, "cannot be mutated anymore");
+    }
+
+    private static boolean keyEquals(@Nullable AsciiString a, CharSequence b) {
+        return a != null && (a == b || a.contentEqualsIgnoreCase(b));
```

Comment:
Do you think `a == b` should be upstreamed into `AsciiString.contentEqualsIgnoreCase` itself?

Reply:
Yes. Please feel free to send a pull request if I don't beat you. :-)